packages/signers/src/__tests__/sign-transaction-test.ts

```diff
@@ -0,0 +1,550 @@
+    });
+});
+
+describe('signAndSendTransactionWithSigners', () => {
+    it('signs and sends the transaction with the provided signers', async () => {
+        expect.assertions(3);
+
+        // Given a transaction with a partial signer A and a sending signer B.
+        const signerA = createMockTransactionPartialSigner('1111' as Address);
+        const signerB = createMockTransactionSendingSigner('2222' as Address);
+        const transaction = createMockTransactionWithSigners([signerA, signerB]);
+
+        // And given signer A and B are mocked to provide the following return values.
+        signerA.signTransactions.mockResolvedValueOnce([{ '1111': '1111_signature' }]);
+        signerB.signAndSendTransactions.mockResolvedValueOnce([new Uint8Array([1, 2, 3])]);
+
+        // When we sign and send this transaction.
+        const transactionSignature = await signAndSendTransactionWithSigners(transaction);
+
+        // Then the sending signer was used to send the transaction.
+        expect(signerA.signTransactions).toHaveBeenCalledWith([transaction]);
+        expect(signerB.signAndSendTransactions).toHaveBeenCalledWith([
+            { ...transaction, signatures: { '1111': '1111_signature' } },
+        ]);
+
+        // And the returned signature matches the one returned by the sending signer.
+        expect(transactionSignature).toStrictEqual(new Uint8Array([1, 2, 3]));
+    });
+
+    it('signs and sends the transaction with extra signers', async () => {
+        expect.assertions(3);
+
+        // Given a partial signer A and a sending signer B with the following mocked return values.
+        const signerA = createMockTransactionPartialSigner('1111' as Address);
+        const signerB = createMockTransactionSendingSigner('2222' as Address);
+        signerA.signTransactions.mockResolvedValueOnce([{ '1111': '1111_signature' }]);
+        signerB.signAndSendTransactions.mockResolvedValueOnce([new Uint8Array([1, 2, 3])]);
+
+        // And given a transaction that only contains signer A in its account metas.
+        const transaction = createMockTransactionWithSigners([signerA]);
+
+        // When we sign and send this transaction whilst providing signer B as an extra signer.
+        const transactionSignature = await signAndSendTransactionWithSigners(transaction, [signerB]);
+
+        // Then both signers were called with the expected parameters.
+        expect(signerA.signTransactions).toHaveBeenCalledWith([transaction]);
+        expect(signerB.signAndSendTransactions).toHaveBeenCalledWith([
+            { ...transaction, signatures: { '1111': '1111_signature' } },
+        ]);
+
+        // And the returned signature matches the one returned by the sending signer.
+        expect(transactionSignature).toStrictEqual(new Uint8Array([1, 2, 3]));
+    });
+
+    it('sends the transaction using the provided fallback sender if no sending signer is used', async () => {
```

Comment:
Because they may not know whether or not a `TransactionSendingSigner` has been provided.

Sure, if the app knows the signers it's using then it can:

```ts
// Sign and send if we know there's a sending signer.
await signAndSendTransactionWithSigners(transaction);

// Sign THEN send if we know there's no sending signer.
const signedTransaction = await signTransactionWithSigners(transaction);
await rpc.sendTransaction(signedTransaction).send();
```

And that code works, as-is.

The reality is we may not know this in advance. For instance, a wallet abstraction such as wallet adapters or the wallet standard may offer a modifying signer or a sending signer or both. This `fallbackSender` enables us to fallback to sending the transaction however we want if no sending signer was identified or provided.

Reply:
I'm happy with that actually. To make sure I'm getting this right, you're saying we should not accept a `fallbackSender` and just fail if no sending signer was identified (ideally with a coded error when we sort all our `// TODO: Coded error.` haha).

Would you not even be tempted to offer an additional helper like the following?

```ts
signAndSendTransactionWithSignersAndFallback(transaction, fallbackSender);
```

Reply:
Resolving this as a decision was made [here](https://github.com/solana-labs/solana-web3.js/pull/1792#issuecomment-1820487421).